services/node_listener.go

```diff
@@ -0,0 +1,291 @@
+		ht.headSubscription.Unsubscribe()
+		ht.headSubscription = nil
+	}
+	if ht.headers != nil {
+		close(ht.headers)
+		ht.headers = nil
+	}
+	ht.Disconnected()
+	return nil
+}
+
+// Updates the latest block number, if indeed the latest, and persists
+// this number in case of reboot. Thread safe.
+func (ht *HeadTracker) Save(n *models.IndexableBlockNumber) error {
+	if n == nil {
+		return errors.New("Cannot save a nil block header")
+	}
+
+	ht.headMutex.Lock()
+	if ht.number == nil || ht.number.ToInt().Cmp(n.ToInt()) < 0 {
+		copy := *n
+		ht.number = &copy
+	}
+	ht.headMutex.Unlock()
+	return ht.store.Save(n)
+}
+
+// Returns the latest block header being tracked, or nil.
+func (ht *HeadTracker) Get() *models.IndexableBlockNumber {
+	ht.headMutex.RLock()
+	defer ht.headMutex.RUnlock()
+	return ht.number
+}
+
+func (ht *HeadTracker) Attach(t HeadTrackable) string {
+	ht.trackersMutex.Lock()
+	defer ht.trackersMutex.Unlock()
+	id := uuid.Must(uuid.NewV4()).String()
+	ht.trackers[id] = t
+	if ht.connected {
+		t.Connected()
+	}
+	return id
+}
+
+func (ht *HeadTracker) Detach(id string) {
+	ht.trackersMutex.Lock()
+	defer ht.trackersMutex.Unlock()
+	t, present := ht.trackers[id]
+	if ht.connected && present {
+		t.Disconnected()
+	}
+	delete(ht.trackers, id)
+}
+
+func (ht *HeadTracker) IsConnected() bool { return ht.connected }
+
+func (ht *HeadTracker) Connected() {
+	ht.trackersMutex.RLock()
+	defer ht.trackersMutex.RUnlock()
+	ht.connected = true
+	for _, t := range ht.trackers {
+		logger.WarnIf(t.Connected())
+	}
+}
+
+func (ht *HeadTracker) Disconnected() {
+	ht.trackersMutex.RLock()
+	defer ht.trackersMutex.RUnlock()
+	ht.connected = false
+	for _, t := range ht.trackers {
+		t.Disconnected()
+	}
+}
+
+func (ht *HeadTracker) OnNewHead(head *models.BlockHeader) {
+	ht.trackersMutex.RLock()
+	defer ht.trackersMutex.RUnlock()
+	for _, t := range ht.trackers {
+		t.OnNewHead(head)
+	}
+}
+
+func (ht *HeadTracker) subscribeToNewHeads() (models.EthSubscription, error) {
+	sub, err := ht.store.TxManager.SubscribeToNewHeads(ht.headers)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		err := <-sub.Err()
+		if err != nil {
+			logger.Warnw("Error in new head subscription, disconnected", "err", err)
+			ht.Stop()
+			ht.reconnectLoop()
+		}
+	}()
+	return sub, nil
+}
+
+func (ht *HeadTracker) listenToNewHeads() {
+	if ht.number != nil {
+		logger.Info("Tracking logs from block ", ht.number.FriendlyString(), " with hash ", ht.number.Hash.String())
+	}
+	for header := range ht.headers {
+		number := header.IndexableBlockNumber()
+		logger.Debugw(fmt.Sprintf("Received header %v", number.FriendlyString()), "hash", header.Hash())
+		if err := ht.Save(number); err != nil {
+			logger.Error(err.Error())
+		} else {
+			ht.OnNewHead(&header)
+		}
+	}
+}
+
+func (ht *HeadTracker) reconnectLoop() {
+	ht.sleeper.Reset()
+	for {
+		logger.Info("Reconnecting to node ", ht.store.Config.EthereumURL, " in ", ht.sleeper.Duration())
+		ht.sleeper.Sleep()
```

Comment:
Should we reinstantiate `ht.sleeper` on successful start? The scenario I'm thinking of is if a connection is dropped once, and the backoff time reaches a minute, and then connection is dropped again later the backoff timer will start at one minute, which might be too slow. Or maybe there's something I'm missing that would catch that?

Reply:
We pass the `Sleeper` in on tests so that we don't need to wait to make an assertion. If we always reinstantiated it, we would need to refactor what we pass in to be a `SleeperFactory`. I felt like it was overkill at the time, especially considering the size of the PR. But you let me know if you still want to change it.